<?xml version="1.0"?>
<robot name="cleaning_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  <!-- Base link - center of the robot chassis -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.30 0.25 0.10"/>
      </geometry>
      <material name="blue">
        <color rgba="0.2 0.2 0.8 1.0"/>
      </material>
    </visual>
  </link>

  <!-- Camera link - RealSense D455 mounting position -->
  <!-- Adjust these values based on your actual camera mounting:
       - x: forward/backward (positive = forward)
       - y: left/right (positive = left)
       - z: up/down (positive = up)
       - rpy: roll/pitch/yaw rotation
  -->
  <joint name="camera_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_link"/>
    <!-- Default: camera mounted 15cm forward, at same height as base center -->
    <origin xyz="0.15 0.0 0.0" rpy="0 0 0"/>
  </joint>

  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.026 0.090 0.025"/>
      </geometry>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
  </link>

  <!-- RealSense publishes optical frames, which have a different orientation -->
  <!-- Optical frame: Z forward, X right, Y down -->
  <joint name="camera_depth_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_depth_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  
  <link name="camera_depth_frame"/>

  <joint name="camera_depth_optical_joint" type="fixed">
    <parent link="camera_depth_frame"/>
    <child link="camera_depth_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <link name="camera_depth_optical_frame"/>

  <joint name="camera_color_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_color_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  
  <link name="camera_color_frame"/>

  <joint name="camera_color_optical_joint" type="fixed">
    <parent link="camera_color_frame"/>
    <child link="camera_color_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <link name="camera_color_optical_frame"/>

  <joint name="camera_infra1_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_infra1_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  
  <link name="camera_infra1_frame"/>

  <joint name="camera_infra1_optical_joint" type="fixed">
    <parent link="camera_infra1_frame"/>
    <child link="camera_infra1_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <link name="camera_infra1_optical_frame"/>

  <joint name="camera_infra2_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_infra2_frame"/>
    <origin xyz="0 0.055 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  
  <link name="camera_infra2_frame"/>

  <joint name="camera_infra2_optical_joint" type="fixed">
    <parent link="camera_infra2_frame"/>
    <child link="camera_infra2_optical_frame"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <link name="camera_infra2_optical_frame"/>

  <joint name="camera_imu_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_imu_optical_frame"/>
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>
  
  <link name="camera_imu_optical_frame"/>

</robot>
